### ISSUE TITLE
[utilities.spacing] improve list definitions and let vertical/horizontal directions

### ISSUE THREAD

**Issue by hayatbiralem** [inuitcss/inuitcss#228]
Hi,

## First

I think `$inuit-spacing-directions` and `$inuit-spacing-properties` are nice but their use is faulty.

For example usage of size variable is right:

```scss
// variable
$inuit-spacing-sizes: (
  null: $inuit-global-spacing-unit,
  '-tiny': $inuit-global-spacing-unit-tiny,
  '-small': $inuit-global-spacing-unit-small,
  '-large': $inuit-global-spacing-unit-large,
  '-huge': $inuit-global-spacing-unit-huge,
  '-none': 0
) !default;

// usage
@each $size, $value in $inuit-spacing-sizes {
    // ...
}

// apparently usage
@each $size-namespace, $size-value in $inuit-spacing-sizes {
    // ...
}
```

But usages of property and direction are wrong:

```scss
// variable
$inuit-spacing-properties: (
  'padding': 'padding',
  'margin': 'margin',
) !default;

// usage
@each $property, $property-namespace in $inuit-spacing-properties {
  // ...
}
```

`$property-namespace` variable is in the wrong place. It must be this way:

```scss
// usage
@each $property-namespace, $property in $inuit-spacing-properties {
  // ...
}
```

**I think we should do this improvement.**
First part is done.

## Second

If we use it like this then we can use lists as value. So our new option comes at this point. We can define vertical and horizontal directions in the `$inuit-spacing-directions`.

I know that, this approach is optional and probably should not be in the framework but I think we should be able easily if we want it.

So, I wrote some code to make it clear.

SassMeister example [here](http://www.sassmeister.com/gist/736d4004c9d5acd57582a499ae235e54).

```scss
/* ==========================================================================
   #SPACING
   ========================================================================== */

/**
 * Utility classes to put specific spacing values onto elements. The below loop
 * will generate us a suite of classes like:
 *
 *   .u-margin-top {}
 *   .u-padding-left-large {}
 *   .u-margin-right-small {}
 *   .u-padding {}
 *   .u-padding-right-none {}
 */

/* stylelint-disable string-quotes */

$inuit-spacing-properties: (
  'padding': 'padding',
  'margin': 'margin',
) !default;

$inuit-spacing-directions: (
  null: null,
  '-top'       : '-top',
  '-right'     : '-right',
  '-bottom'    : '-bottom',
  '-left'      : '-left',
  '-horizontal': '-left' '-right', // optional but possible, use it if you want
  '-vertical'  : '-top' '-bottom', // optional but possible, use it if you want
) !default;

$inuit-spacing-sizes: (
  null: $inuit-global-spacing-unit,
  '-tiny': $inuit-global-spacing-unit-tiny,
  '-small': $inuit-global-spacing-unit-small,
  '-large': $inuit-global-spacing-unit-large,
  '-huge': $inuit-global-spacing-unit-huge,
  '-none': 0
) !default;


@each $property-namespace, $property in $inuit-spacing-properties {

  @each $direction-namespace, $direction in $inuit-spacing-directions {
    
    @each $size-namespace, $value in $inuit-spacing-sizes {

      .u-#{$property-namespace}#{$direction-namespace}#{$size-namespace} {
        @if type-of($direction) == 'list' {
          @each $sub-direction in $direction {
            #{$property}#{$sub-direction}: $value !important;
          }
        } @else {
          #{$property}#{$direction}: $value !important;
        }
      }

    }
    
  }

}
```

Result:

```scss
.u-padding-horizontal {
  padding-left: 24px !important;
  padding-right: 24px !important;
}

.u-padding-horizontal-tiny {
  padding-left: 6px !important;
  padding-right: 6px !important;
}

.u-padding-horizontal-small {
  padding-left: 12px !important;
  padding-right: 12px !important;
}

.u-padding-horizontal-large {
  padding-left: 48px !important;
  padding-right: 48px !important;
}

.u-padding-horizontal-huge {
  padding-left: 96px !important;
  padding-right: 96px !important;
}

.u-padding-horizontal-none {
  padding-left: 0 !important;
  padding-right: 0 !important;
}

.u-padding-vertical {
  padding-top: 24px !important;
  padding-bottom: 24px !important;
}

.u-padding-vertical-tiny {
  padding-top: 6px !important;
  padding-bottom: 6px !important;
}

.u-padding-vertical-small {
  padding-top: 12px !important;
  padding-bottom: 12px !important;
}

.u-padding-vertical-large {
  padding-top: 48px !important;
  padding-bottom: 48px !important;
}

.u-padding-vertical-huge {
  padding-top: 96px !important;
  padding-bottom: 96px !important;
}

.u-padding-vertical-none {
  padding-top: 0 !important;
  padding-bottom: 0 !important;
}
```

Second part is also done here :)

So what do you think about this?

✌️

**Comment by herzinger** [comment]
>`$property-namespace` variable is in the wrong place. It must be this way: [...]

@hayatbiralem You are wrong, and I really can't see why would you think otherwise, really. 

The convention, following the usage of `foreach` loops or similar functions in most languages that have it, is `key, value`, which is how it is today. Your version works just the same, but there is no reason to make this change, and the most conventional order is the current one.

The only improvement I see in your code is the addition of "horizontal" and "vertical" directions. I use them myself. But you can achieve it easily by predefining the correct object in a by project level, so...

**Comment by hayatbiralem** [comment]
> You are wrong, and I really can't see why you would think otherwise, really.

@herzinger  If am I wrong, is the following code correct?

```scss
$inuit-spacing-directions: (
  null: null,
  '-top'          : '-top',
  '-right'        : '-right',
  '-bottom'       : '-bottom',
  '-left'         : '-left',
  '-left' '-right': '-horizontal',
  '-top' '-bottom': '-vertical',
) !default;
```

Lists can not be a key but namespaces can so namespaces must always be used as key in my opinion. I think this approach is the best practise.

Also css declaration order is the same:

```scss
.any {
  padding#{$direction-namespace}: $size;
  padding-left: 24px;
  padding-vertical: 24px;
}
```

Direction namespace came first and value came after.

You are right, partly, it can be used in two ways if only string values are used, but it is useful to develop a standard in my opinion.

**Comment by herzinger** [comment]
@hayatbiralem I see, your way makes it easier to set multiple rules to some directions. 
Okay, I maintain that under normal circunstancies, the current order is the best convention, and that you were wrong in saying it was wrong at the start (it's not), but if there's a plausible reason (there is), I see no problem in changing the direction. I stand corrected, LGTM.

Allowing easier special directions classes is the merit in this change, nothing else. But it's a good merit.

As far as I'm concerned, I'm convinced. You should make a pull request.

**Comment by hayatbiralem** [comment]
@herzinger I see your point too and I'm sorry if I could not explain it correctly. Thanks, by the way, I'll make a pull request 👍

**Comment by anselmh** [comment]
I think we can close this issue now since it landed in a002df332af9066a3791a7309277c9f9b8025931, right @hayatbiralem? If so, please close :)

**Comment by hayatbiralem** [comment]
@anselmh It's a pleasure for me 😄
